nvflare/app_common/streamers/container_streamer.py

```
@@ -0,0 +1,255 @@
+            raise ValueError(error)
+
+        last = shareable.get(_KEY_LAST)
+        if last:
+            # Check if all entries are added
+            if self.size != len(self.container):
+                err = f"Container size {len(self.container)} does not match expected size {self.size}"
+                self.logger.error(err)
+                raise ValueError(err)
+            else:
+                stream_ctx[_CTX_RESULT] = self.container
+                return False, make_reply(ReturnCode.OK)
+        else:
+            # continue streaming
+            return True, make_reply(ReturnCode.OK)
+
+    def finalize(self, stream_ctx: StreamContext, fl_ctx: FLContext):
+        self.logger.debug(f"Container streaming is done for container type {type(self.container)}")
+
+
+class _EntryConsumerFactory(ConsumerFactory):
+
+    def get_consumer(self, stream_ctx: StreamContext, fl_ctx: FLContext) -> ObjectConsumer:
+        return _EntryConsumer(stream_ctx)
+
+
+class _EntryProducer(ObjectProducer):
+    def __init__(self, container, entry_timeout):
+        self.logger = get_obj_logger(self)
+        if not container:
+            error = "Can't stream empty container"
+            self.logger.error(error)
+            raise ValueError(error)
+
+        self.container = container
+        if isinstance(container, dict):
+            self.iterator = iter(container.items())
+        else:
+            self.iterator = iter(container)
+        self.size = len(container)
+        self.count = 0
+        self.last = False
+        self.entry_timeout = entry_timeout
+
+    def produce(
+        self,
+        stream_ctx: StreamContext,
+        fl_ctx: FLContext,
+    ) -> Tuple[Shareable, float]:
+
+        try:
+            entry = next(self.iterator)
+            self.count += 1
+            self.last = self.count >= self.size
+        except StopIteration:
+            self.logger.error(f"Producer called too many times {self.count}/{self.size}")
```

This condition can only happen when the container is empty or programming error.
You should just return (None, 0) to ensure aborting, instead of returning a Shareable.

---

Changed.